### PR TITLE
handle inline c-like comments in lexer

### LINKED ISSUE #537: [Feature request] Inline comments
Particularly in longer lines (such as those containing complex macros), it's often useful to add a comment in the middle of a line, like `/*  ... */` in C. This is not currently supported.

I'd personally recommend braces (`{comment goes here}`), since they are currently unused in the language and there doesn't seem to be any use for them.

### PR DESCRIPTION
**Pull request by anderoonies**
fixes #537 

Let me know if there's a more preferable way to do this, I'm not positive the best way to do the peeking and `shiftChars(2)` looks out of place.

### CODE REVIEW

**Review by JL2210 — CHANGES_REQUESTED** [review]

File: src/asm/lexer.c
Comment (anderoonies) [reply src/asm/lexer.c:930]: right, should i handle that with a custom `error` or let it throw syntax error?

Comment (JL2210) [reply src/asm/lexer.c:930]: Neither. You should allow inline comments to span multiple lines.

Comment (anderoonies) [reply src/asm/lexer.c:930]: and if it's unterminated before EOF?

Comment (anderoonies) [reply src/asm/lexer.c:930]: sorry, to clarify—
should i have a custom error message for an inline comment that's unterminated before EOF?

Comment (JL2210) [reply src/asm/lexer.c:930]: That seems reasonable. I would use `error("unterminated comment\n");`

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:925]: To be consistent with other languages, I'd rather they be named "block" comments; they can be used inline, but that's not their typical use.

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:929]: Why 2? Shouldn't it be 1 from the `peek(0)` that led here? (Arguably, I'd rather move that `shiftChars` to the calling site, I think it'd be clearer)

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:937]: This should rather be a warning; it might be legitimate to have these characters in a comment (example: `/* Handle binary ops +/* now */`), and it can still be made into an error if desired (using `-Werror=xxx`). The question then becomes which warning flag to use; I'd suggest `-Wnested-comment`.

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:931]: Arguably, this can become a `switch (nextChar())`. Since that shifts the character being matched, the following `peek`s get an argument of 0, which is better. (You'd move the `shiftChars(1)` at the end of the loop to after it, too.)

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:1526]: This can be simplified to `if (peek(0) == '*')`.

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Hi! Thanks for the contribution.

The structure of the code is good, all I have are some nitpicks, so this should get merged fairly swiftly.

File: src/asm/lexer.c
Comment (anderoonies) [reply src/asm/lexer.c:929]: yes, good catch. i'll move the shiftChars as well, wasn't sure where that was best put

File: src/asm/lexer.c
Comment (anderoonies) [reply src/asm/lexer.c:925]: will make that change in all places 👍

File: src/asm/lexer.c
Comment (JL2210) [inline src/asm/lexer.c:938]: Not "can't" anymore. Maybe "/* in block comment"?

**Review by JL2210 — CHANGES_REQUESTED** [review]

File: src/asm/lexer.c
Comment (anderoonies) [reply src/asm/lexer.c]: warning is pushed up, made the change before seeing your note so it's the verbose `-Wcomment-in-block-comment`. i can update to `-Wnested-comment`, but it's different than `; comment ; comment`, so it might be clearer to call it `block` in the warning

**Comment by anderoonies** [comment]
thanks for the reviews, i'll have updates in about 3 hours

File: src/asm/warning.c
Comment (ISSOtm) [inline src/asm/warning.c:74]: This seems fairly long, especially compared to the `nested-comment` I had suggested earlier. The length is further evidenced through it throwing other warning constants out of alignment, which I have been considering to be a soft cap.

(I'm not fond of the length of e.g. `empty-data-directive`, but too late to turn back)

**Comment by ISSOtm** [comment]
> I think a flag to disable this warning would be useful

https://rgbds.gbdev.io/docs/v0.4.1/rgbasm.1#DIAGNOSTICS

> Note that each of these flag also has a negation (for example, -Wempty-entry enables the warning that -Wno-empty-entry disables).

Those are automatically managed by checking for `no-` at the beginning of the supplied warning flag:

https://github.com/gbdev/rgbds/blob/42b3a173566a70c665066bf9e1f8d2568b9b176c/src/asm/warning.c#L184

Besides, the proposed patch only makes it part of `-Wextra`, which is IMO good.

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:941]: This is buggy, as it will cause `//` to end comments. It would make sense if the `peek(0)` was a `nextChar()`, though.

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:944]: To remove the need for `block_ended`, it'd be simpler to `shiftChars(1)` here, and then return.

File: src/asm/warning.c
Comment (ISSOtm) [inline src/asm/warning.c:39]: Alignment is performed here (and in the header's enum) using tabs.

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:931]: Code style mandates a space after `switch`; you can either [set checkpatch up yourself](https://github.com/gbdev/rgbds/blob/master/CONTRIBUTING.rst#contributing-code), or [look at the CI report](https://github.com/gbdev/rgbds/pull/607/checks) (it only fails on errors, but we prefer avoiding warnings and checks whenever reasonably possible)

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/asm/lexer.c
Comment (ISSOtm) [reply src/asm/lexer.c]: The rationale for that warning is that a block comment starter within a block comment may indicate a runaway block comment (forgot to close it, only closed by a later comment).

Line comments, however, have much more limited scope, and thus cannot really be made runaway :P

**Comment by anderoonies** [comment]
checkpatch is erroring on old diffs, not sure how to resolve that.

**Comment by AntonioND** [comment]
Instead of creating new commits on top of the ones with style errors, squash them with ``git rebase -i``.

File: src/asm/lexer.c
Comment (ISSOtm) [inline src/asm/lexer.c:937]: Actually, this `continue` should be in the other branch: currently, `/* //` ends the comment, but `/* /*/\ doesn't.
Test cases (`weird-block-comment`?) should be added for those, too;

**Review by ISSOtm — CHANGES_REQUESTED** [review]

**Review by ISSOtm — APPROVED** [review]
LGTM!
That said, we shouldn't merge this without updating the documentation. If @anderoonies doesn't want to do it, any of the maintainers can add a commit to the branch, and then we can merge this.

**Comment by anderoonies** [comment]
@ISSOtm i'm happy to, i see docs are sourced from `rgbasm.1` and `rgbasm.5`, is there a more human-readable way to edit those files?

**Comment by JL2210** [comment]
Nope

**Comment by anderoonies** [comment]
😀 ok

**Comment by anderoonies** [comment]
any notes on the copy before i commit?
image

**Comment by JL2210** [comment]
I agree, but I think that's out of the scope of this PR.

**Comment by AntonioND** [comment]
Yeah. I guess that removing that will be pretty easy after this PR is merged

**Comment by ISSOtm** [comment]
Okay, there are some caveats with the man page formatting, but I wasn't aware of them either before bentley told me about them. For example, sentence-ending periods must be at the end of a line, so that Mandoc is able to properly format them. I will no longer annoy @anderoonies about those, though, I'll just make the last fixups myself and merge.

@Rangi42's example is
```asm
    db 1
    db 7
```
btw.

**Pull request merged**
